apps/data-pipeline/library-traffic-scraper/src/lib.ts

```diff
@@ -0,0 +1,232 @@
+  const url = `https://www.lib.uci.edu/sites/all/scripts/occuspace.php?id=${id}`;
+  try {
+    // Double-encoded JSON: double parse required
+    const responseText = await fetch(url).then((r) => r.text());
+    const intermediateJson = JSON.parse(responseText);
+    const parsedResponse =
+      typeof intermediateJson === "string" ? JSON.parse(intermediateJson) : intermediateJson;
+    if (parsedResponse.data && typeof parsedResponse.data === "object")
+      return (parsedResponse as RawRespOK).data;
+    console.warn(`ID ${id} responded with error: ${(parsedResponse as RawRespErr).error}`);
+  } catch (err) {
+    console.error(`Unexpected error while fetching ID ${id}:`, err);
+    throw err;
+  }
+  return null;
+}
+
+/**
+ * Represents the possible statuses for library scraping based on operating hours
+ * - "active": Library location is open — perform scrape every 15 minutes
+ * - "idle": Library is closed — perform scrape every 60 minutes
+ * - "skip": Library is unknown or not applicable for scraping - do not scrape
+ */
+type ScrapeStatus = "active" | "idle" | "skip";
+
+export function getScrapeStatus(library: "LL" | "SL" | "LGSC"): ScrapeStatus {
+  const pacificTime = new Date(
+    new Date().toLocaleString("en-US", { timeZone: "America/Los_Angeles" }),
+  );
+  const hour = pacificTime.getHours();
+  const day = pacificTime.getDay();
+
+  // Langson Library (LL) and Science Library (SL) - same operating hours
+  if (library === "LL" || library === "SL") {
+    // Mon–Thu: Open from 8am to 11pm
+    if (day >= 1 && day <= 4 && hour >= 8 && hour < 23) return "active";
+    // Fri: Open from 8am to 6pm
+    if (day === 5 && hour >= 8 && hour < 18) return "active";
+    // Sat: Open from 1pm to 5pm
+    if (day === 6 && hour >= 13 && hour < 17) return "active";
+    // Sun: Open from 1pm to 9pm
+    if (day === 0 && hour >= 13 && hour < 21) return "active";
+
+    // Outside of above hours, the library is closed
+    return "idle";
+  }
+
+  // Libraries Gateway Study Center (LGSC)
+  if (library === "LGSC") {
+    // Mon–Thu: Open from 8am until 3am the next day
+    if (day >= 1 && day <= 4 && (hour >= 8 || hour < 3)) return "active";
+    // Fri: Open from 8am to 9pm
+    if (day === 5 && hour >= 8 && hour < 21) return "active";
+    // Sat: Open from 5pm to 9pm
+    if (day === 6 && hour >= 17 && hour < 21) return "active";
+    // Sun: Open from 5pm to 3am the next day
+    if (day === 0 && (hour >= 17 || hour < 3)) return "active";
+
+    // Outside of above hours, the library is closed
+    return "idle";
+  }
+  return "skip";
+}
+
+export async function doScrape(db: ReturnType<typeof database>) {
+  console.log("Starting library traffic scrape.");
+  const locationMeta = await collectLocationMeta();
+
+  const currentTime = new Date();
+  const currentMinute = currentTime.getMinutes();
+
+  for (const [id, meta] of Object.entries(locationMeta)) {
+    const libraryCodeMap = {
```

Comment:
this does not need to be instantiated every time we enter this loop and should be declared once at file scope